File: linda-logic/src/main/java/it/unibo/coordination/linda/logic/LogicTuple.java
package it.unibo.coordination.linda.logic;

import alice.tuprolog.Struct;
import alice.tuprolog.Term;
import alice.tuprolog.Var;
import it.unibo.coordination.linda.core.Tuple;

import java.util.Objects;

public interface LogicTuple extends Tuple, Comparable<LogicTuple> {

    static LogicTuple of(String format, Object... args) {
        return LogicTuple.of(Term.createTerm(Objects.requireNonNull(String.format(format, args))));
    }

    static LogicTuple of(Term term) {
        return new LogicTupleImpl(term);
    }

    static Struct getPattern() {
        return Struct.of("tuple", Var.of("T"));
    }

    static Struct getPattern(Term term) {
        return Struct.of("tuple", Objects.requireNonNull(term));
    }

    Struct asTerm();

    Term getTuple();

    default LogicTemplate toTemplate() {
        return LogicTemplate.of(getTuple());
    }

    @Override
    boolean equals(Object o);

    @Override
    int hashCode();

    default int compareTo(LogicTuple o) {
        return getTuple().toString().compareTo(o.getTuple().toString());
    }

    static boolean equals(LogicTuple t1, LogicTuple t2) {
        if (t1 == t2) return true;
        if (t1 == null || t2 == null) return false;
        return Objects.equals(t1.asTerm(), t2.asTerm());
    }

    static int hashCode(LogicTuple t) {
        return Objects.hashCode(t.asTerm().toString());
    }
}

File: linda-logic/src/main/kotlin/it/unibo/coordination/linda/logic/LogicTuple.kt
package it.unibo.coordination.linda.logic

import it.unibo.coordination.linda.core.Tuple
import it.unibo.tuprolog.core.Struct
import it.unibo.tuprolog.core.Term
import it.unibo.tuprolog.core.Var
import it.unibo.tuprolog.core.parsing.parse
import java.util.*

interface LogicTuple : Tuple<LogicTuple>, Comparable<LogicTuple> {

    override val value: Term

    fun asTerm(): Struct

    override fun equals(other: Any?): Boolean

    override fun hashCode(): Int

    @JvmDefault
    override fun compareTo(other: LogicTuple): Int = value.toString().compareTo(other.value.toString())

    companion object {

        @JvmStatic
        fun of(tuple: String): LogicTuple = of(Term.parse(tuple))

        @JvmStatic
        fun of(term: Term): LogicTuple = LogicTupleImpl(term)

        @JvmStatic
        val pattern: Struct = Struct.of("tuple", Var.of("T"))

        @JvmStatic
        fun getPattern(term: Term): Struct = Struct.of("tuple", term)

        @JvmStatic
        fun equals(t1: LogicTuple?, t2: LogicTuple?): Boolean =
                if (t1 === t2) true
                else if (t1 == null || t2 == null) false
                else t1.value == t2.value

        @JvmStatic
        fun hashCode(t: LogicTuple): Int = Objects.hashCode(t.asTerm())
    }
}